tests/Aspire.Hosting.Tests/MSBuildTests.cs

```diff
@@ -0,0 +1,136 @@
+
+public class MSBuildTests
+{
+    /// <summary>
+    /// Tests that when an AppHost has a ProjectReference to a library project, a warning is emitted.
+    /// </summary>
+    [Fact]
+    public void EnsureWarningsAreEmittedWhenProjectReferencingLibraries()
+    {
+        var repoRoot = GetRepoRoot();
+        var tempDirectory = Directory.CreateTempSubdirectory("AspireHostingTests");
+        try
+        {
+            var libraryDirectory = Path.Combine(tempDirectory.FullName, "Library");
+            Directory.CreateDirectory(libraryDirectory);
+
+            File.WriteAllText(Path.Combine(libraryDirectory, "Library.csproj"), """
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+</Project>
+""");
+            File.WriteAllText(Path.Combine(libraryDirectory, "Class1.cs"), """
+namespace Library;
+
+public class Class1
+{
+}
+""");
+
+            var appHostDirectory = Path.Combine(tempDirectory.FullName, "AppHost");
+            Directory.CreateDirectory(appHostDirectory);
+
+            File.WriteAllText(Path.Combine(appHostDirectory, "AppHost.csproj"), $"""
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+    <IsAspireHost>true</IsAspireHost>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="{repoRoot}\src\Aspire.Hosting\Aspire.Hosting.csproj" AspireHostLibrary="true" />
+
+    <ProjectReference Include="..\Library\Library.csproj" />
+  </ItemGroup>
+
+</Project>
+""");
+            File.WriteAllText(Path.Combine(appHostDirectory, "Program.cs"), """
+var builder = DistributedApplication.CreateBuilder();
+builder.Build().Run();
+""");
+
+            File.WriteAllText(Path.Combine(appHostDirectory, "Directory.Build.props"), $"""
+<Project>
+  <Import Project="{repoRoot}\src\Aspire.Hosting\build\Aspire.Hosting.props" />
+</Project>
+""");
+            File.WriteAllText(Path.Combine(appHostDirectory, "Directory.Build.targets"), $"""
+<Project>
+  <PropertyGroup>
+    <_SuppressImportAspireHostingOrchestrationTargets>true</_SuppressImportAspireHostingOrchestrationTargets>
+  </PropertyGroup>
+  <Import Project="{repoRoot}\src\Aspire.Hosting\build\Aspire.Hosting.targets" />
+  <Import Project="{repoRoot}\src\Aspire.Hosting.Sdk\SDK\Sdk.targets" />
+</Project>
+""");
+
+            var output = new StringBuilder();
+            var outputDone = new ManualResetEvent(false);
+            var process = new Process();
+            // set '-nodereuse:false -p:UseSharedCompilation=false' so the MSBuild and Roslyn server processes don't hang around, which may hang the test in CI
+            process.StartInfo = new ProcessStartInfo("dotnet", $"build -nodereuse:false -p:UseSharedCompilation=false")
```

Comment:
The only thing that worries me from this is how dotnet resolution might differ in different machines and CI. I wonder if we should get .dotnet by analyzing the CurrentProcess arg0, so that we use the same dotnet as the one that was launched for running these tests.

Reply:
For running tests on the build machine setting that or explicitly using `dotnet` from `.dotnet` should be enough.
On helix there will only be the `dotnet` with version matching the one in `global.json`, so it shouldn't be an issue there.

Reply:
Local runs can do the same thing - compute a path to `.dotnet/dotnet`, and use that.

Reply:
> or explicitly using dotnet from .dotnet should be enough.

I don't think we can rely on this. If the  machine already has the SDK listed in global.json globally installed, then the build will skip restoring it and creating a .dotnet folder, and instead will just use the global one. We've been bit by this in the past.

Reply:
Good point. I guess in that case use `dotnet` from `PATH`, which then should be the correct version?

Reply:
@eerhardt can correct me if I'm wrong, but I believe you can get to the one running the tests via Process.GetCurrentProcess() and then inspecting arg0

Reply:
I think that only works when you set it on the ProcessStartInfo in the first place and used that to start the process?

alternative: Process.GetCurrentProcess().MainModule.FileName;

Reply:
These approaches don't work because the process being run is:

> C:\git\aspire\artifacts\bin\Aspire.Hosting.Tests\Debug\net8.0\testhost.exe

We are already getting the "repo root" in this test. So the only viable options I see are:

1. Doing what I'm doing here and using the `dotnet` on the path.
2. Using `{repoRoot}\.dotnet\dotnet`.

Thoughts on those 2 options? (or are there others?)

Reply:
You can't depend on (2) because of the reason @joperezr mentioned where your system dotnet version matches the one in global.json, thus you don't get any `.dotnet`.

One possible thing is to add `$(DotnetTool)` from msbuild to `PATH` when running the tests, which should make the expected `dotnet` available in `PATH.

Reply:
> One possible thing is to add $(DotnetTool) from msbuild to PATH when running the tests

How would one do that? I'm running the tests in VS using the Test Explorer.

Either way, this wouldn't change the test code - we would just invoke `dotnet` and let the PATH take care of it.

Reply:
forgot about VS! 😬